core/internal/viewstate/view-state_test.go

```diff
@@ -0,0 +1,283 @@
+				release()
+			}
+		}
+		if c.wait {
+			ok, release := s.WaitAndHoldView(view)
+			require.Equal(t, c.ok, ok, c.desc)
+			if ok {
+				release()
+			}
+		}
+		if c.active {
+			activeView, release := s.WaitAndHoldActiveView()
+			require.Equal(t, view, activeView, c.desc)
+			release()
+		}
+	}
+}
+
+func TestConcurrent(t *testing.T) {
+	const nrConcurrent = 7
+	const nrViewChanges = 17
+
+	wg := new(sync.WaitGroup)
+
+	runConcurrently := func(f func()) {
+		wg.Add(1)
+
+		go func() {
+			f()
+			wg.Done()
+		}()
+	}
+
+	var (
+		active    uint64
+		requested uint64
+		started   uint64
+		finished  uint64
+	)
+
+	s := New()
+
+	assertActiveView := func(view uint64) {
+		lastActive := atomic.SwapUint64(&active, view)
+		assert.True(t, lastActive <= view,
+			"Active view number cannot decrease")
+		assert.True(t, started <= view,
+			"View change cannot begin before view released")
+		assert.True(t, finished == view,
+			"View change cannot complete before view released")
+	}
+
+	waitAndHoldActiveView := func(view uint64) {
+		for {
+			activeView, release := s.WaitAndHoldActiveView()
+			assertActiveView(activeView)
+			go release()
+			if activeView >= view {
+				break
+			}
+		}
+	}
+
+	waitAndHoldView := func(view uint64) {
+		ok, release := s.WaitAndHoldView(view)
+		if !ok {
+			return
+		}
+		assertActiveView(view)
+		go release()
+	}
+
+	requestViewChange := func(view uint64) {
+		lastActive := atomic.LoadUint64(&active)
+		lastRequested := atomic.LoadUint64(&requested)
+		lastFinished := atomic.LoadUint64(&finished)
+
+		if !s.RequestViewChange(view) {
+			return
+		}
+		atomic.CompareAndSwapUint64(&requested, lastRequested, view)
+		assert.True(t, lastRequested < view,
+			"Requested view change number cannot decrease")
+		assert.True(t, lastActive < view,
+			"Requested view change number must be higher than last active")
+		assert.True(t, lastFinished < view,
+			"Requested view change number must be higher than last finished")
+
+	}
+
+	startViewChange := func(view uint64) {
+		ok, release := s.StartViewChange(view)
+		if !ok {
+			return
+		}
+		assert.True(t, started < view,
+			"Started view change number cannot decrease")
+		started = view
+		assert.True(t, finished < view,
+			"Started view change number must be higher than last finished")
+		assert.True(t, active < view,
+			"Started view change number must be higher than last active")
+		go release()
+	}
+
+	finishViewChange := func(view uint64) {
+		ok, release := s.FinishViewChange(view)
+		if !ok {
+			return
+		}
+		assert.True(t, finished < view,
+			"Finished view change number cannot decrease")
+		atomic.StoreUint64(&finished, view)
+		assert.True(t, started <= view,
+			"Finished view change number cannot be lower than last started")
+		assert.True(t, active < view,
+			"Finished view change number must be higher than last active")
+		go release()
+	}
+
+	for i := 0; i < nrConcurrent; i++ {
+		for view := uint64(0); view <= nrViewChanges; view++ {
+			view := view
```

Comment:
This is a common idiom in Go. It is typically used when using a function literal (lambda/anonymous function) to start a goroutine from a loop. This line creates a new independent variable `view` for each iteration of the loop. The new variable captures the current value of the loop variable `view` and shadows it. Then any following goroutine started from in the loop will use that new variable. Otherwise, the started goroutine would use the original variable that keep asynchronously changing as the loop continues to iterate.

Reply:
Thank you for explanation. I understand.

I find that this is FAQ. Sorry to bother you.
https://golang.org/doc/faq#closures_and_goroutines